### ISSUE TITLE
unsupported operand type in eventlet\semaphore.py (greenpool)

### ISSUE THREAD

**Issue by MPagel** [eventlet/eventlet#364]
```
Traceback (most recent call last):
  File "D:\Programs\Python\Python27\lib\site-packages\eventlet\hubs\hub.py", line 457, in fire_timers
    timer()
  File "D:\Programs\Python\Python27\lib\site-packages\eventlet\hubs\timer.py", line 58, in __call__
    cb(*args, **kw)
  File "D:\Programs\Python\Python27\lib\site-packages\eventlet\greenpool.py", line 141, in _do_map
    gi.spawn(func, *args)
  File "D:\Programs\Python\Python27\lib\site-packages\eventlet\greenpool.py", line 207, in spawn
    gt = self.pool.spawn(func, *args, **kw)
  File "D:\Programs\Python\Python27\lib\site-packages\eventlet\greenpool.py", line 71, in spawn
    self.sem.acquire()
  File "D:\Programs\Python\Python27\lib\site-packages\eventlet\semaphore.py", line 123, in acquire
    self.counter -= 1
TypeError: unsupported operand type(s) for -=: 'str' and 'int'

$ pip install eventlet
Requirement already satisfied: eventlet in d:\programs\python\python27\lib\site-packages
Requirement already satisfied: greenlet>=0.3 in d:\programs\python\python27\lib\site-packages (from eventlet)
Requirement already satisfied: enum-compat in d:\programs\python\python27\lib\site-packages (from eventlet)
Requirement already satisfied: enum34 in d:\programs\python\python27\lib\site-packages (from enum-compat->eventlet)

$ pip show python
Name: eventlet
Version: 0.20.0
Summary: Highly concurrent networking library
Home-page: http://eventlet.net
Author: Linden Lab
Author-email: eventletdev@lists.secondlife.com
License: UNKNOWN
Location: d:\programs\python\python27\lib\site-packages
Requires: greenlet, enum-compat
```

Python 2.7.11 (v2.7.11:6d1b6a68f775, Dec  5 2015, 20:32:19) [MSC v.1500 32 bit (Intel)] on win32

**Comment by temoto** [comment]
Can you share code to reproduce issue?

**Comment by temoto** [comment]
I think you have something like `pool.resize('2')`. Python doesn't automagically convert between numbers and strings.

**Comment by MPagel** [comment]
did a bit of further debugging.  It's because my call to establish the pool was GreenPool(size=cfg.get('GLOBAL','Max_AWS_Connections')) where cfg is an object of type ConfigParser.RawConfigParser().  So, it's sending it in as a string.  So, chalk this up to (mostly) end-user error.  However, I recommend the following check in semaphore.py:

```
    def __init__(self, value=1):
        try:
            te = False if isinstance(value, (int, long)) else True
            self.counter = long(value)
            te_text = ""
        except TypeError:
            te = True
            te_text = "Type error in increment of counter. counter is now type {}. {} was type {}\n".format(type(self.counter),value,type(value))
        if ((value <= 0) or te):
            raise ValueError(te_text+"Semaphore must be initialized with a positive "
                             "number, got %s" % value)
        self._waiters = collections.deque()
```

**Comment by MPagel** [comment]
(note, this blocks the behavior of a GreenPool of size 0)

**Comment by temoto** [comment]
`te = not isinstance(...)` ;-)

**Comment by MPagel** [comment]
Thanks. Hopefully it should be Pagel-proof now :P 

`te = not isinstance(...)` - um, yeah.  /facepalm.  You certainly have a better implementation now than copying `value` and then doing a `try: counter += 0` as my first work-around was

**Comment by temoto** [comment]
The fix is merged in master 79292bd so it will be included in emergency bugfix release 0.20.1 soon. Thanks and have great time.
